### PR TITLE
[GSK-2540, GSK-2558, GSK-2663] RAG toolset

### PR DESCRIPTION
**Pull request by pierlj**
- Add the correctness evaluator and associated tests. 
- Add a generator class `KnowledgeBaseTestsetGenerator` in the `giskard/rag` module. 
- Add utilities to build a knowledge base inside the generator. This include `VectorStore`, `Document` and `OpenAIEmbeddings` classes. These are highly inspired from `langchain`, but much simpler. 
- Add tests for all these classes.

### CODE REVIEW

File: giskard/llm/evaluators/base.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ def _make_evaluate_prompt(self, model: BaseModel, input_vars, model_output, row_
             model_output=model_output,
         )
 
-    def _make_evaluate_functions(self, model: BaseModel, input_vars, model_output):
+    def _make_evaluate_functions(self):
```
Comment (mattbit) [inline giskard/llm/evaluators/base.py:81]: I wouldn't touch this

File: giskard/llm/evaluators/correctness.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    REFERENCE_ANSWER = 1
+    REFERENCE_CONTEXT = 2
+
+
+class CorrectnessEvaluator(LLMBasedEvaluator):
+    _default_eval_prompt = CORRECTNESS_EVALUATION_PROMPT
+    _required_features = ["question", "reference_answer"]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Comment (mattbit) [inline giskard/llm/evaluators/correctness.py:51]: not needed

**Review by mattbit — COMMENTED** [review]
Looks good, thanks! I will try it out locally before merging.

File: giskard/llm/evaluators/correctness.py (hunk 2 above)
Comment (pierlj) [reply giskard/llm/evaluators/correctness.py:51]: Indeed

File: giskard/rag/__init__.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+from .knowledge_base_testset_generator import KnowledgeBaseTestsetGenerator
+
+__all__ = [KnowledgeBaseTestsetGenerator]
```
Comment (mattbit) [inline giskard/rag/__init__.py:3]:
```suggestion
__all__ = ["KnowledgeBaseTestsetGenerator"]
```

File: giskard/rag/knowledge_base_testset_generator.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        prompt = self._prevent_context_window_overflow(prompt)
+        return self._llm_complete(prompt, self._make_generate_input_functions("question"))
+
+    def _generate_answer_from_context(self, question, context):
+        prompt = self._answer_generation_prompt.format(question=question, context=context)
+        prompt = self._prevent_context_window_overflow(prompt)
+        return self._llm_complete(prompt, self._make_generate_input_functions("answer"))
+
+    def _extract_seed_context(self):
+        seed_context = np.random.choice(self.knowledge_base.documents)
```
[the remark this answers is not in the document]
Comment (mattbit) [reply giskard/rag/knowledge_base_testset_generator.py:82]: Can be good to get a proper generator with `np.random.default_rng()`

File: giskard/rag/knowledge_base_testset_generator.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    def __init__(
+        self,
+        knowledge_df,
+        model_name: str,
+        model_description: str,
+        context_neighbors: int = 4,
+        context_similarity_threshold: float = 0.2,
+        context_window_length: int = 8192,
+        embedding_model: EmbeddingsBase = None,
+        language: str = "english",
```
Comment (mattbit) [inline giskard/rag/knowledge_base_testset_generator.py:28]: Not clear what code you are using for language. Maybe check IETF BCP 47? It's common to use "en", "fr", etc.

File: pyproject.toml [hunk 6]
```diff
@@ -174,6 +174,7 @@ dependencies = [
     "markdown",                 # needed for display of scan results in notebook
     "colorama",                 # needed for the scan
     "griffe>=0.36.9",
+    "faiss-cpu>=1.7.4",
```
Comment (mattbit) [inline pyproject.toml:177]: this not good 😢

Comment (pierlj) [reply pyproject.toml:177]: What's wrong with that ?

File: giskard/rag/knowledge_base_testset_generator.py (hunk 5 above)
Comment (pierlj) [reply giskard/rag/knowledge_base_testset_generator.py:28]: Yes let's keep [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1) as we did for [GSK-2152].

**Comment by mattbit** [comment]
Pushed some fixes

File: docs/index.md [hunk 7]
```diff
@@ -16,6 +16,7 @@ open_source/installation_library/index
 open_source/scan/index
 open_source/customize_tests/index
 open_source/integrate_tests/index
+open_source/testset_generation/index
```
Comment (luca-martial) [inline docs/index.md:19]: Could you put it right after scan given its importance?

File: docs/open_source/index.md [hunk 8]
```diff
@@ -8,6 +8,7 @@ installation_library/index
 scan/index
 customize_tests/index
 integrate_tests/index
+testset_generation/index
```
Comment (luca-martial) [inline docs/open_source/index.md:11]: Same here

**Review by luca-martial — CHANGES_REQUESTED** [review]

File: docs/index.md (hunk 7 above)
Comment (pierlj) [reply docs/index.md:19]: Sure!

**Review by mattbit — APPROVED** [review]

**Review by luca-martial — APPROVED** [review]

**Pull request merged**